### PR TITLE
BC: Separated regex matching in NodeNameResolver->isName()

### PR DESCRIPTION
**Pull request by staabm**
as discussed in https://github.com/rectorphp/rector-src/pull/4950#discussion_r1320547797

----

Afte todays patches rectorphp is running 100 seconds faster over the [Mautic](https://twitter.com/Mautic) codebase.

latest stable release takes ~ 625.33s
rector@dev takes ~ 523s on my machine

.. and requires way less RAM

### CODE REVIEW

File: rules/Php55/Rector/String_/StringClassNameToClassConstantRector.php [hunk 1]
```diff
@@ -176,7 +176,7 @@ private function shouldSkip(string $classLikeName): bool
         }
 
         foreach ($this->classesToSkip as $classToSkip) {
-            if ($this->nodeNameResolver->isStringName($classLikeName, $classToSkip)) {
+            if ($this->nodeNameResolver->matchesStringName($classLikeName, $classToSkip)) {
```
Comment (staabm) [inline rules/Php55/Rector/String_/StringClassNameToClassConstantRector.php:183]: this requires a wildcard in the pattern now. should we call regular isStringName() beforehand to keep BC?

File: rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php [hunk 2]
```diff
@@ -142,7 +142,7 @@ private function shouldSkip(Class_ $class): bool
         if ($this->transformOnNamespaces !== []) {
             $className = (string) $this->nodeNameResolver->getName($class);
             foreach ($this->transformOnNamespaces as $transformOnNamespace) {
-                if (! $this->nodeNameResolver->isStringName($className, $transformOnNamespace)) {
+                if (! $this->nodeNameResolver->matchesStringName($className, $transformOnNamespace)) {
```
Comment (staabm) [inline rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php:149]: this requires a wildcard in the pattern now. should we call regular isStringName() beforehand to keep BC?

File: rules/Php55/Rector/String_/StringClassNameToClassConstantRector.php (hunk 1 above)
Comment (TomasVotruba) [reply rules/Php55/Rector/String_/StringClassNameToClassConstantRector.php:183]: Yeah, this should allow both types

File: rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php (hunk 2 above)
Comment (TomasVotruba) [reply rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php:149]: Yes, this should check both :+1:

File: packages/NodeNameResolver/NodeNameResolver.php [hunk 3]
```diff
@@ -198,27 +193,18 @@ public function isStringName(string $resolvedName, string $desiredName): bool
...
-                return StringUtils::isMatch($resolvedName, $desiredName);
-            }
+        return strtolower($resolvedName) === strtolower($desiredName);
+    }
 
-            // is probably fnmatch
-            if (\str_contains($desiredName, '*')) {
-                return fnmatch($desiredName, $resolvedName, FNM_NOESCAPE);
-            }
+    public function matchesStringName(string $resolvedName, string $desiredNamePattern): bool
```
Comment (TomasVotruba) [inline packages/NodeNameResolver/NodeNameResolver.php]: The `$resolvedName` should accept both string and identifier IMO.

File: rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php (hunk 2 above)
Comment (staabm) [reply rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php:149]: done

File: packages/NodeNameResolver/NodeNameResolver.php (hunk 3 above)
Comment (staabm) [reply packages/NodeNameResolver/NodeNameResolver.php]: done

**Comment by samsonasik** [comment]
@staabm please rebase to fix conflict, thank you.

**Comment by TomasVotruba** [comment]
@staabm rector-symfony needs one last check, that name is not `Expr`. Then its ready to go :+1:

**Comment by TomasVotruba** [comment]
Thanks :+1: 

@staabm How are we doing on the performance numbers? :)

**Comment by staabm** [comment]
updated the PRs description above

**Comment by TomasVotruba** [comment]
That's awesome improvement :clap: :clap: :clap: 

 I wonder if we use regex by default somewhere else but actually don't need it :thinking:

**Comment by samsonasik** [comment]
@staabm it seems cause error on rector-phpunit:

- https://github.com/rectorphp/rector-phpunit/pull/251

**Pull request merged**